File: apps/mobile/src/lib/points.ts
// Point reads (RN port of apps/web/src/lib/points.ts). Uses the Go API when
// configured, else Supabase RPCs. Returns @safecity/shared PointSummary.
import type { Category, FeatureValue, PointSummary } from '@safecity/shared';
import { api, ApiError, apiEnabled, qs } from './api';
import { supabase } from './supabase';

interface RpcRow {
  id: string;
  name: string;
  category: PointSummary['category'];
  address: string | null;
  description?: string | null;
  photos?: string[] | null;
  lng: number;
  lat: number;
  verify_status: PointSummary['verifyStatus'];
  distance_m?: number;
  features: PointSummary['features'];
}

function mapRow(r: RpcRow): PointSummary {
  return {
    id: r.id,
    name: r.name,
    category: r.category,
    address: r.address,
    description: r.description ?? null,
    photos: r.photos ?? [],
    lng: r.lng,
    lat: r.lat,
    verifyStatus: r.verify_status,
    distanceM: r.distance_m ?? undefined,
    features: r.features ?? {},
  };
}

/** Points within radius (m) of a location, nearest first. */
export async function pointsNear(lng: number, lat: number, radiusM = 1500): Promise<PointSummary[]> {
  if (apiEnabled) {
    return api.get<PointSummary[]>(`/points/near${qs({ lng, lat, radius: radiusM })}`);
  }
  const { data, error } = await supabase.rpc('points_near', { lng, lat, radius_m: radiusM });
  if (error) throw error;
  return ((data ?? []) as RpcRow[]).map(mapRow);
}

/** Points whose geometry falls inside a map bounding box. */
export async function pointsInBbox(
  minLng: number,
  minLat: number,
  maxLng: number,
  maxLat: number,
): Promise<PointSummary[]> {
  if (apiEnabled) {
    return api.get<PointSummary[]>(
      `/points/bbox${qs({ min_lng: minLng, min_lat: minLat, max_lng: maxLng, max_lat: maxLat })}`,
    );
  }
  const { data, error } = await supabase.rpc('points_in_bbox', {
    min_lng: minLng,
    min_lat: minLat,
    max_lng: maxLng,
    max_lat: maxLat,
  });
  if (error) throw error;
  return ((data ?? []) as RpcRow[]).map(mapRow);
}

/** A single point by id (coords + feature values), or null if not found. */
export async function pointById(id: string): Promise<PointSummary | null> {
  if (apiEnabled) {
    try {
      return await api.get<PointSummary>(`/points/${id}`);
    } catch (e) {
      if (e instanceof ApiError && e.status === 404) return null;
      throw e;
    }
  }
  const { data, error } = await supabase.rpc('point_detail', { p_id: id });
  if (error) throw error;
  const r = ((data ?? []) as RpcRow[])[0];
  return r ? mapRow(r) : null;
}

export interface NewPoint {
  name: string;
  category: Category;
  lng: number;
  lat: number;
  address?: string;
  description?: string;
  features: Record<string, FeatureValue>;
  photos?: string[];
}

/** Create a point. Returns the new id. Uses the Go API when configured, else add_point RPC. */
export async function addPoint(input: NewPoint): Promise<string> {
  if (apiEnabled) {
    const res = await api.post<{ id: string }>('/points', {
      name: input.name,
      category: input.category,
      lat: input.lat,
      lng: input.lng,
      address: input.address ?? '',
      description: input.description ?? '',
      features: input.features,
      photos: input.photos ?? [],
    });
    return res.id;
  }
  const { data, error } = await supabase.rpc('add_point', {
    p_name: input.name,
    p_category: input.category,
    p_lng: input.lng,
    p_lat: input.lat,
    p_address: input.address || null,
    p_description: input.description || null,
    p_features: input.features,
    p_photos: input.photos ?? [],
  });
  if (error) throw error;
  return data as string;
}

export interface PointHit {
  id: string;
  name: string;
  category: Category;
  address: string | null;
}

/** Name/address search across all points (Supabase-direct; no API endpoint yet). */
export async function searchPointsByName(query: string, limit = 6): Promise<PointHit[]> {
  const q = query.trim();
  if (q.length < 2) return [];
  const pattern = `%${q}%`;
  const { data, error } = await supabase
    .from('points')
    .select('id, name, category, address')
    .or(`name.ilike.${pattern},address.ilike.${pattern}`)
    .limit(limit);
  if (error) throw error;
  return (data ?? []) as PointHit[];
}

File: apps/api/internal/store/points.go
package store

import (
	"context"
	"errors"

	"github.com/jackc/pgx/v5"
)

// PointSummary is a map/list point with its feature values. Field names match the
// web's @safecity/shared PointSummary so the rating engine consumes it unchanged.
type PointSummary struct {
	ID                 string            `json:"id"`
	Name               string            `json:"name"`
	Category           string            `json:"category"`
	Address            *string           `json:"address"`
	Lng                float64           `json:"lng"`
	Lat                float64           `json:"lat"`
	VerifyStatus       string            `json:"verifyStatus"`
	DistanceM          *float64          `json:"distanceM,omitempty"`
	Features           map[string]string `json:"features"`
	IsBusiness         bool              `json:"isBusiness"`
	VerifiedPaid       bool              `json:"verifiedPaid"`
	SubscriptionActive bool              `json:"subscriptionActive"`
}

// PointDetail adds description + photos for the detail view.
type PointDetail struct {
	ID                 string            `json:"id"`
	Name               string            `json:"name"`
	Category           string            `json:"category"`
	Address            *string           `json:"address"`
	Description        *string           `json:"description"`
	Photos             []string          `json:"photos"`
	Lng                float64           `json:"lng"`
	Lat                float64           `json:"lat"`
	VerifyStatus       string            `json:"verifyStatus"`
	Features           map[string]string `json:"features"`
	IsBusiness         bool              `json:"isBusiness"`
	VerifiedPaid       bool              `json:"verifiedPaid"`
	SubscriptionActive bool              `json:"subscriptionActive"`
}

// Reads are public (guest-first; RLS read_all). They reuse the existing PostGIS
// RPCs and cast uuid/enum columns to text so they scan into plain Go strings.

const pointsNearSQL = `
select id::text, name, category::text, address, lng, lat, verify_status::text, distance_m, features,
       is_business, verified_paid, subscription_active
from points_near($1, $2, $3)`

const pointsInBBoxSQL = `
select id::text, name, category::text, address, lng, lat, verify_status::text, features,
       is_business, verified_paid, subscription_active
from points_in_bbox($1, $2, $3, $4)`

const pointDetailSQL = `
select id::text, name, category::text, address, description, photos, lng, lat, verify_status::text, features,
       is_business, verified_paid, subscription_active
from point_detail($1)`

// PointsNear returns points within radiusM metres of (lng,lat), nearest first.
func (s *Store) PointsNear(ctx context.Context, lng, lat, radiusM float64) ([]PointSummary, error) {
	rows, err := s.db.Pool.Query(ctx, pointsNearSQL, lng, lat, radiusM)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	out := []PointSummary{}
	for rows.Next() {
		var p PointSummary
		if err := rows.Scan(&p.ID, &p.Name, &p.Category, &p.Address,
			&p.Lng, &p.Lat, &p.VerifyStatus, &p.DistanceM, &p.Features,
			&p.IsBusiness, &p.VerifiedPaid, &p.SubscriptionActive); err != nil {
			return nil, err
		}
		ensureFeatures(&p.Features)
		out = append(out, p)
	}
	return out, rows.Err()
}

// PointsInBBox returns points whose geometry intersects the bounding box.
func (s *Store) PointsInBBox(ctx context.Context, minLng, minLat, maxLng, maxLat float64) ([]PointSummary, error) {
	rows, err := s.db.Pool.Query(ctx, pointsInBBoxSQL, minLng, minLat, maxLng, maxLat)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	out := []PointSummary{}
	for rows.Next() {
		var p PointSummary
		if err := rows.Scan(&p.ID, &p.Name, &p.Category, &p.Address,
			&p.Lng, &p.Lat, &p.VerifyStatus, &p.Features,
			&p.IsBusiness, &p.VerifiedPaid, &p.SubscriptionActive); err != nil {
			return nil, err
		}
		ensureFeatures(&p.Features)
		out = append(out, p)
	}
	return out, rows.Err()
}

// PointDetail returns a single point, or (nil, nil) if not found.
func (s *Store) PointDetail(ctx context.Context, id string) (*PointDetail, error) {
	var p PointDetail
	err := s.db.Pool.QueryRow(ctx, pointDetailSQL, id).Scan(
		&p.ID, &p.Name, &p.Category, &p.Address, &p.Description, &p.Photos,
		&p.Lng, &p.Lat, &p.VerifyStatus, &p.Features,
		&p.IsBusiness, &p.VerifiedPaid, &p.SubscriptionActive)
	if errors.Is(err, pgx.ErrNoRows) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	ensureFeatures(&p.Features)
	if p.Photos == nil {
		p.Photos = []string{}
	}
	return &p, nil
}

// PointHit is a lightweight name/address search match.
type PointHit struct {
	ID       string  `json:"id"`
	Name     string  `json:"name"`
	Category string  `json:"category"`
	Address  *string `json:"address"`
}

// Boosts points owned by an active business account to the top of search results,
// keeping the existing name/address filter. Scoped to text search only (not /near
// or /bbox) — the user asked for search-result priority specifically; map-browsing
// order is a separate, larger decision if ever wanted later.
const searchPointsSQL = `
select p.id::text, p.name, p.category::text, p.address
from points p
where p.name ilike $1 or p.address ilike $1
order by is_business_user(p.created_by) desc, p.name
limit $2`

// SearchPoints matches points by name or address substring (case-insensitive).
func (s *Store) SearchPoints(ctx context.Context, query string, limit int) ([]PointHit, error) {
	pattern := "%" + query + "%"
	rows, err := s.db.Pool.Query(ctx, searchPointsSQL, pattern, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	out := []PointHit{}
	for rows.Next() {
		var h PointHit
		if err := rows.Scan(&h.ID, &h.Name, &h.Category, &h.Address); err != nil {
			return nil, err
		}
		out = append(out, h)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}
	// Best-effort analytics: every returned point counts as one search appearance.
	if len(out) > 0 {
		ids := make([]string, len(out))
		for i := range out {
			ids[i] = out[i].ID
		}
		go func() {
			_, _ = s.db.Pool.Exec(context.Background(),
				`update points set search_appearances = search_appearances + 1 where id = any($1::uuid[])`, ids)
		}()
	}
	return out, nil
}

func ensureFeatures(m *map[string]string) {
	if *m == nil {
		*m = map[string]string{}
	}
}
